compiler/packages/babel-plugin-react-compiler/src/HIR/visitors.ts

```diff
@@ -1149,3 +1153,72 @@ export function* eachTerminalOperand(terminal: Terminal): Iterable<Place> {
     }
   }
 }
+
+/**
+ * Helper class for traversing scope blocks in HIR-form.
+ */
+export class ScopeBlockTraversal {
+  // Live stack of active scopes
+  #activeScopes: Array<ScopeId> = [];
+  blockInfos: Map<
+    BlockId,
+    | {
+        kind: 'end';
+        scope: ReactiveScope;
+        pruned: boolean;
+      }
+    | {
+        kind: 'begin';
```

Comment:
A single block can only be either the start of a scope block or the fallthrough. This is because `BuildReactiveScopeTerminals` inserts a scope start block and a scope fallthrough block for every scope. So by construction, a block can be either the start or fallthrough but not both.

```
// before BuildReactiveScopeTerminals
bb0:
[0]
[1]  -┐
[2]   | scope@0
[3]  -┘
...

// after
bb0:
  [0]
  ScopeTerminal block=bb1 fallthr=bb2

bb1: <-- scope begin block
  [1]
  [2]
  Goto bb2

bb2: <-- scope end block
  [3]
  [4]
```